File: core/src/main/java/com/graphhopper/routing/util/parsers/helpers/OSMValueExtractor.java
package com.graphhopper.routing.util.parsers.helpers;

import com.graphhopper.reader.ReaderWay;
import com.graphhopper.routing.ev.DecimalEncodedValue;
import com.graphhopper.routing.ev.EdgeIntAccess;
import com.graphhopper.routing.ev.MaxSpeed;
import com.graphhopper.util.DistanceCalcEarth;
import com.graphhopper.util.Helper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.regex.Pattern;

import static com.graphhopper.util.Helper.toLowerCase;

public class OSMValueExtractor {

    private static final Pattern TON_PATTERN = Pattern.compile("tons?");
    private static final Pattern MGW_PATTERN = Pattern.compile("mgw");
    private static final Pattern WSPACE_PATTERN = Pattern.compile("\\s");
    private static final Pattern METER_PATTERN = Pattern.compile("meters?|mtrs?|mt|m\\.");
    private static final Pattern INCH_PATTERN = Pattern.compile("\"|\'\'");
    private static final Pattern FEET_PATTERN = Pattern.compile("\'|feet");
    private static final Pattern APPROX_PATTERN = Pattern.compile("~|approx");
    private static final Logger logger = LoggerFactory.getLogger(OSMValueExtractor.class);

    private OSMValueExtractor() {
        // utility class
    }

    public static void extractTons(int edgeId, EdgeIntAccess edgeIntAccess, ReaderWay way, DecimalEncodedValue valueEncoder, List<String> keys) {
        final String rawValue = way.getFirstPriorityTag(keys);
        double value = stringToTons(rawValue);

        if (Double.isNaN(value)) value = Double.POSITIVE_INFINITY;

        valueEncoder.setDecimal(false, edgeId, edgeIntAccess, value);
        // too many
//        if (value - valueEncoder.getDecimal(false, edgeFlags) > 2)
//            logger.warn("Value " + value + " for " + valueEncoder.getName() + " was too large and truncated to " + valueEncoder.getDecimal(false, edgeFlags));
    }

    /**
     * This parses the weight for a conditional value like "delivery @ (weight > 7.5)"
     */
    public static double conditionalWeightToTons(String value) {
        try {
            int index = value.indexOf("weight>"); // maxweight or weight
            if (index < 0) {
                index = value.indexOf("weight >");
                if (index > 0) index += "weight >".length();
            } else {
                index += "weight>".length();
            }
            if (index > 0) {
                int lastIndex = value.indexOf(')', index); // (value) or value
                if (lastIndex < 0) lastIndex = value.length() - 1;
                if(lastIndex > index) return OSMValueExtractor.stringToTons(value.substring(index, lastIndex));
            }
            return Double.NaN;
        } catch (Exception ex) {
            throw new RuntimeException("value " + value, ex);
        }
    }

    public static double stringToTons(String value) {
        value = TON_PATTERN.matcher(toLowerCase(value)).replaceAll("t");
        value = MGW_PATTERN.matcher(value).replaceAll("").trim();
        if (isInvalidValue(value))
            return Double.NaN;

        double factor = 1;
        if (value.endsWith("st")) {
            value = value.substring(0, value.length() - 2);
            factor = 0.907194048807;
        } else if (value.endsWith("t")) {
            value = value.substring(0, value.length() - 1);
        } else if (value.endsWith("lbs")) {
            value = value.substring(0, value.length() - 3);
            factor = 0.00045359237;
        } else if (value.endsWith("kg")) {
            value = value.substring(0, value.length() - 2);
            factor = 0.001;
        }

        try {
            return Double.parseDouble(value) * factor;
        } catch (NumberFormatException e) {
            return Double.NaN;
        }
    }

    public static void extractMeter(int edgeId, EdgeIntAccess edgeIntAccess, ReaderWay way, DecimalEncodedValue valueEncoder, List<String> keys) {
        final String rawValue = way.getFirstPriorityTag(keys);
        double value = stringToMeter(rawValue);

        if (Double.isNaN(value)) value = Double.POSITIVE_INFINITY;

        valueEncoder.setDecimal(false, edgeId, edgeIntAccess, value);
        // too many
//        if (value - valueEncoder.getDecimal(false, edgeFlags) > 2)
//            logger.warn("Value " + value + " for " + valueEncoder.getName() + " was too large and truncated to " + valueEncoder.getDecimal(false, edgeFlags));
    }

    public static double stringToMeter(String value) {
        value = WSPACE_PATTERN.matcher(toLowerCase(value)).replaceAll("");
        value = METER_PATTERN.matcher(value).replaceAll("m");
        value = INCH_PATTERN.matcher(value).replaceAll("in");
        value = FEET_PATTERN.matcher(value).replaceAll("ft");
        if (isInvalidValue(value))
            return Double.NaN;
        double factor = 1;
        double offset = 0;
        if (value.startsWith("~") || value.contains("approx")) {
            value = APPROX_PATTERN.matcher(value).replaceAll("").trim();
            factor = 0.8;
        }

        if (value.endsWith("in")) {
            int startIndex = value.indexOf("ft");
            String inchValue;
            if (startIndex < 0) {
                startIndex = 0;
            } else {
                startIndex += 2;
            }

            inchValue = value.substring(startIndex, value.length() - 2);
            value = value.substring(0, startIndex);
            try {
                offset = Double.parseDouble(inchValue) * 0.0254;
            } catch (NumberFormatException e) {
                return Double.NaN;
            }
        }

        if (value.endsWith("ft")) {
            value = value.substring(0, value.length() - 2);
            factor *= 0.3048;
        } else if (value.endsWith("cm")) {
            value = value.substring(0, value.length() - 2);
            factor *= 0.01;
        } else if (value.endsWith("m")) {
            value = value.substring(0, value.length() - 1);
        }

        if (value.isEmpty()) {
            return offset;
        }

        try {
            return Double.parseDouble(value) * factor + offset;
        } catch (NumberFormatException e) {
            return Double.NaN;
        }
    }

    public static boolean isInvalidValue(String value) {
        value = toLowerCase(value);
        return value.isEmpty() || value.startsWith("default") || value.equals("none") || value.equals("unknown")
                || value.contains("unrestricted") || value.startsWith("〜")
                || value.contains("narrow") || value.equals("unsigned") || value.equals("fixme") || value.equals("small")
                || value.contains(";") || value.contains(":") || value.contains("(")
                || value.contains(">") || value.contains("<") || value.contains("-")
                // only support '.' and no German decimals
                || value.contains(",");
    }

    /**
     * @return the speed in km/h
     */
    public static double stringToKmh(String str) {
        if (Helper.isEmpty(str))
            return Double.NaN;

        // on some German autobahns and a very few other places
        if ("none".equals(str))
            return MaxSpeed.UNLIMITED_SIGN_SPEED;

        if (str.endsWith(":rural") || str.endsWith(":trunk"))
            return 80;

        if (str.endsWith(":urban"))
            return 50;

        if (str.equals("walk") || str.endsWith(":living_street"))
            return 6;

        int mpInteger = str.indexOf("mp");
        int knotInteger = str.indexOf("knots");
        int kmInteger = str.indexOf("km");
        int kphInteger = str.indexOf("kph");

        double factor;
        if (mpInteger > 0) {
            str = str.substring(0, mpInteger).trim();
            factor = DistanceCalcEarth.KM_MILE;
        } else if (knotInteger > 0) {
            str = str.substring(0, knotInteger).trim();
            factor = 1.852; // see https://en.wikipedia.org/wiki/Knot_%28unit%29#Definitions
        } else {
            if (kmInteger > 0) {
                str = str.substring(0, kmInteger).trim();
            } else if (kphInteger > 0) {
                str = str.substring(0, kphInteger).trim();
            }
            factor = 1;
        }

        double value;
        try {
            value = Integer.parseInt(str) * factor;
        } catch (Exception ex) {
            return Double.NaN;
        }

        if (value <= 0) {
            return Double.NaN;
        }

        return value;
    }
}

File: core/src/main/kotlin/com/graphhopper/routing/util/parsers/helpers/OSMValueExtractor.kt
package com.graphhopper.routing.util.parsers.helpers

import com.graphhopper.reader.ReaderWay
import com.graphhopper.routing.ev.DecimalEncodedValue
import com.graphhopper.routing.ev.EdgeIntAccess
import com.graphhopper.util.Helper.toLowerCase
import java.util.regex.Pattern

object OSMValueExtractor {

    private val TON_PATTERN = Pattern.compile("tons?")
    private val MGW_PATTERN = Pattern.compile("mgw")
    private val WSPACE_PATTERN = Pattern.compile("\\s")
    private val METER_PATTERN = Pattern.compile("meters?|mtrs?|mt|m\\.")
    private val INCH_PATTERN = Pattern.compile("\"|''")
    private val FEET_PATTERN = Pattern.compile("'|feet")
    private val APPROX_PATTERN = Pattern.compile("~|approx")

    @JvmStatic
    fun extractTons(edgeId: Int, edgeIntAccess: EdgeIntAccess, way: ReaderWay, valueEncoder: DecimalEncodedValue, keys: List<String>) {
        val rawValue = way.getFirstValue(keys)
        var value = stringToTons(rawValue)

        if (value.isNaN()) value = Double.POSITIVE_INFINITY

        valueEncoder.setDecimal(false, edgeId, edgeIntAccess, value)
    }

    /**
     * This parses the weight for a conditional value like "delivery @ (weight > 7.5)"
     */
    @JvmStatic
    fun conditionalWeightToTons(value: String): Double {
        try {
            var index = value.indexOf("weight>") // maxweight or weight
            if (index < 0) {
                index = value.indexOf("weight >")
                if (index > 0) index += "weight >".length
            } else {
                index += "weight>".length
            }
            if (index > 0) {
                var lastIndex = value.indexOf(')', index) // (value) or value
                if (lastIndex < 0) lastIndex = value.length
                if (lastIndex > index)
                    return stringToTons(value.substring(index, lastIndex))
            }
            return Double.NaN
        } catch (ex: Exception) {
            throw RuntimeException("value $value", ex)
        }
    }

    @JvmStatic
    fun stringToTons(value: String): Double {
        var v = TON_PATTERN.matcher(toLowerCase(value)).replaceAll("t")
        v = MGW_PATTERN.matcher(v).replaceAll("").trim()
        if (isInvalidValue(v))
            return Double.NaN

        var factor = 1.0
        if (v.endsWith("st")) {
            v = v.substring(0, v.length - 2)
            factor = 0.907194048807
        } else if (v.endsWith("t")) {
            v = v.substring(0, v.length - 1)
        } else if (v.endsWith("lbs")) {
            v = v.substring(0, v.length - 3)
            factor = 0.00045359237
        } else if (v.endsWith("kg")) {
            v = v.substring(0, v.length - 2)
            factor = 0.001
        }

        return try {
            v.toDouble() * factor
        } catch (e: NumberFormatException) {
            Double.NaN
        }
    }

    @JvmStatic
    fun extractMeter(edgeId: Int, edgeIntAccess: EdgeIntAccess, way: ReaderWay, valueEncoder: DecimalEncodedValue, keys: List<String>) {
        val rawValue = way.getFirstValue(keys)
        var value = stringToMeter(rawValue)

        if (value.isNaN()) value = Double.POSITIVE_INFINITY

        valueEncoder.setDecimal(false, edgeId, edgeIntAccess, value)
    }

    @JvmStatic
    fun stringToMeter(value: String): Double {
        var v = WSPACE_PATTERN.matcher(toLowerCase(value)).replaceAll("")
        v = METER_PATTERN.matcher(v).replaceAll("m")
        v = INCH_PATTERN.matcher(v).replaceAll("in")
        v = FEET_PATTERN.matcher(v).replaceAll("ft")
        if (isInvalidValue(v))
            return Double.NaN
        var factor = 1.0
        var offset = 0.0
        if (v.startsWith("~") || v.contains("approx")) {
            v = APPROX_PATTERN.matcher(v).replaceAll("").trim()
            factor = 0.8
        }

        if (v.endsWith("in")) {
            var startIndex = v.indexOf("ft")
            if (startIndex < 0) {
                startIndex = 0
            } else {
                startIndex += 2
            }

            val inchValue = v.substring(startIndex, v.length - 2)
            v = v.substring(0, startIndex)
            try {
                offset = inchValue.toDouble() * 0.0254
            } catch (e: NumberFormatException) {
                return Double.NaN
            }
        }

        if (v.endsWith("ft")) {
            v = v.substring(0, v.length - 2)
            factor *= 0.3048
        } else if (v.endsWith("cm")) {
            v = v.substring(0, v.length - 2)
            factor *= 0.01
        } else if (v.endsWith("m")) {
            v = v.substring(0, v.length - 1)
        }

        if (v.isEmpty()) {
            return offset
        }

        return try {
            v.toDouble() * factor + offset
        } catch (e: NumberFormatException) {
            Double.NaN
        }
    }

    @JvmStatic
    fun isInvalidValue(value: String): Boolean {
        val v = toLowerCase(value)
        return v.isEmpty() || v.startsWith("default") || v == "none" || v == "unknown"
                || v.contains("unrestricted") || v.startsWith("〜")
                || v.contains("narrow") || v == "unsigned" || v == "fixme" || v == "small"
                || v.contains(";") || v.contains(":") || v.contains("(")
                || v.contains(">") || v.contains("<") || v.contains("-")
                // only support '.' and no German decimals
                || v.contains(",")
    }
}
